File: utils/validators.py
"""
Input Validation Helpers
Reusable validation functions for API inputs
Includes security sanitization to prevent NoSQL injection and XSS attacks
"""
import math
import re
from config import MAX_SEARCH_RADIUS, MAX_SCOOTER_ID_LENGTH, MIN_PASSWORD_LENGTH

# ===========================================
# SECURITY: Patterns to detect malicious input
# ===========================================

# MongoDB operator injection patterns
MONGODB_OPERATORS = re.compile(r'\$[a-zA-Z]+')  # Matches $gt, $ne, $where, etc.

# Common injection patterns
INJECTION_PATTERNS = [
    r'\$where',           # JavaScript execution
    r'\$regex',           # Regex DoS
    r'\$expr',            # Aggregation expressions
    r'\$function',        # Server-side JavaScript
    r'{\s*["\']?\$',      # JSON with MongoDB operators
    r'\[\s*{',            # Array of objects (potential injection)
]

# HTML/Script injection patterns (XSS)
XSS_PATTERNS = [
    r'<script',
    r'javascript:',
    r'on\w+\s*=',         # onclick=, onerror=, etc.
    r'<iframe',
    r'<object',
    r'<embed',
]

# Compiled patterns for efficiency
INJECTION_REGEX = re.compile('|'.join(INJECTION_PATTERNS), re.IGNORECASE)
XSS_REGEX = re.compile('|'.join(XSS_PATTERNS), re.IGNORECASE)


def sanitize_string(value, field_name="input"):
    """
    Sanitize a string input to prevent injection attacks
    Returns: (is_safe: bool, sanitized_value: str or error_message: str)
    """
    if value is None:
        return True, None
    
    # Must be a string (not dict, list, etc.)
    if not isinstance(value, str):
        # If it's a dict or list, it could be an injection attempt
        if isinstance(value, (dict, list)):
            return False, f"{field_name} contains invalid data structure (possible injection attempt)"
        # Try to convert to string
        value = str(value)
    
    # Check for MongoDB operator injection
    if MONGODB_OPERATORS.search(value):
        return False, f"{field_name} contains invalid characters ($ operator not allowed)"
    
    # Check for common injection patterns
    if INJECTION_REGEX.search(value):
        return False, f"{field_name} contains potentially malicious content"
    
    # Check for XSS patterns
    if XSS_REGEX.search(value):
        return False, f"{field_name} contains invalid HTML/script content"
    
    return True, value


def sanitize_input(data, field_name="input"):
    """
    Sanitize any input (handles strings, dicts, lists)
    For dicts/lists from JSON body, ensures no MongoDB operators are present
    """
    if data is None:
        return True, None
    
    if isinstance(data, str):
        return sanitize_string(data, field_name)
    
    if isinstance(data, (int, float)):
        # Numbers are safe, but check for special values
        if isinstance(data, float) and (math.isnan(data) or math.isinf(data)):
            return False, f"{field_name} contains invalid number (NaN or Infinity)"
        return True, data
    
    if isinstance(data, dict):
        # Check all keys and values in dict
        for key, value in data.items():
            # Keys should not contain $ (MongoDB operators)
            if isinstance(key, str) and key.startswith('$'):
                return False, f"Invalid field name in {field_name} ($ prefix not allowed)"
            
            # Recursively check values
            is_safe, result = sanitize_input(value, f"{field_name}.{key}")
            if not is_safe:
                return False, result
        return True, data
    
    if isinstance(data, list):
        # Check all items in list
        for i, item in enumerate(data):
            is_safe, result = sanitize_input(item, f"{field_name}[{i}]")
            if not is_safe:
                return False, result
        return True, data
    
    # For other types, convert to string and check
    return sanitize_string(str(data), field_name)

# US Geographic Bounds (Continental US + some buffer for border areas)
US_BOUNDS = {
    'lat_min': 24.396308,   # Southern tip of Florida Keys
    'lat_max': 49.384358,   # Northern border with Canada
    'lng_min': -125.0,      # West coast (Washington)
    'lng_max': -66.93457    # East coast (Maine)
}

# Extended bounds including Alaska, Hawaii, Puerto Rico
EXTENDED_US_BOUNDS = {
    'lat_min': 17.5,        # Puerto Rico / US Virgin Islands
    'lat_max': 71.5,        # Northern Alaska
    'lng_min': -180.0,      # Alaska (crosses date line)
    'lng_max': -65.0        # Puerto Rico
}


def validate_coordinates(lat, lng, check_us_bounds=False, allow_null_island=False):
    """
    Validate latitude and longitude values with comprehensive checks
    
    Args:
        lat: Latitude value (string or number)
        lng: Longitude value (string or number)
        check_us_bounds: If True, validates coordinates are within US territory
        allow_null_island: If True, allows (0, 0) coordinates
    
    Returns: (is_valid: bool, result: tuple(lat, lng) or error_message: str)
    """
    # Check for None/empty values
    if lat is None or lng is None:
        return False, "Latitude and longitude are required"
    
    if lat == '' or lng == '':
        return False, "Latitude and longitude cannot be empty"
    
    # Try to convert to float
    try:
        lat = float(lat)
        lng = float(lng)
    except (ValueError, TypeError):
        return False, "Coordinates must be valid numbers (e.g., 30.2672, -97.7431)"
    
    # Check for NaN or Infinity
    if math.isnan(lat) or math.isnan(lng):
        return False, "Coordinates cannot be NaN (Not a Number)"
    
    if math.isinf(lat) or math.isinf(lng):
        return False, "Coordinates cannot be infinite"
    
    # Basic range validation
    if not (-90 <= lat <= 90):
        return False, f"Latitude must be between -90 and 90 degrees (got {lat})"
    
    if not (-180 <= lng <= 180):
        return False, f"Longitude must be between -180 and 180 degrees (got {lng})"
    
    # Check for Null Island (0, 0) - common data entry error
    if not allow_null_island and lat == 0 and lng == 0:
        return False, "Coordinates (0, 0) appear to be invalid - this is 'Null Island' in the Atlantic Ocean"
    
    # Optional: Check if within US bounds
    if check_us_bounds:
        is_valid, error = _check_us_bounds(lat, lng)
        if not is_valid:
            return False, error
    
    # Round to reasonable precision (6 decimal places = ~11cm accuracy)
    lat = round(lat, 6)
    lng = round(lng, 6)
    
    return True, (lat, lng)


def _check_us_bounds(lat, lng):
    """
    Check if coordinates are within US territory (including Alaska, Hawaii, Puerto Rico)
    Returns: (is_valid: bool, error_message: str or None)
    """
    # Check Continental US first
    if (US_BOUNDS['lat_min'] <= lat <= US_BOUNDS['lat_max'] and 
        US_BOUNDS['lng_min'] <= lng <= US_BOUNDS['lng_max']):
        return True, None
    
    # Check Hawaii (roughly)
    if 18.5 <= lat <= 22.5 and -160.5 <= lng <= -154.5:
        return True, None
    
    # Check Alaska (roughly - complex due to Aleutian Islands crossing date line)
    if 51.0 <= lat <= 71.5:
        # Main Alaska
        if -180.0 <= lng <= -129.0:
            return True, None
        # Aleutian Islands (wrap around)
        if 170.0 <= lng <= 180.0:
            return True, None
    
    # Check Puerto Rico / US Virgin Islands
    if 17.5 <= lat <= 18.6 and -68.0 <= lng <= -64.5:
        return True, None
    
    return False, f"Coordinates ({lat}, {lng}) appear to be outside US territory. Please verify the location."


def validate_coordinates_strict(lat, lng):
    """
    Strict coordinate validation for scooter placement (US bounds enforced)
    """
    return validate_coordinates(lat, lng, check_us_bounds=True, allow_null_island=False)


def get_coordinate_suggestions(lat, lng):
    """
    Provide helpful suggestions for coordinate issues
    Returns a list of suggestion strings
    """
    suggestions = []
    
    try:
        lat = float(lat)
        lng = float(lng)
    except (ValueError, TypeError):
        suggestions.append("Make sure coordinates are numbers, not text")
        suggestions.append("Example format: 30.2672 for latitude, -97.7431 for longitude")
        return suggestions
    
    # Check if lat/lng might be swapped (common error)
    if -180 <= lat <= 180 and -90 <= lng <= 90:
        if abs(lat) > 90 or abs(lng) > 180:
            suggestions.append("Latitude and longitude may be swapped. Latitude should be between -90 and 90.")
    
    # Check for missing negative sign on longitude (common for US)
    if 66 <= lng <= 125:
        suggestions.append(f"US longitudes are typically negative. Did you mean -{lng}?")
    
    # Check for common city coordinate mistakes
    if lat == 0 and lng == 0:
        suggestions.append("(0, 0) is 'Null Island' in the Atlantic. Enter actual coordinates.")
    
    return suggestions


def validate_radius(radius):
    """
    Validate search radius
    Returns: (is_valid: bool, result: float or error_message: str)
    """
    try:
        radius = float(radius)
    except (ValueError, TypeError):
        return False, "Radius must be a valid number"
    
    if radius <= 0:
        return False, "Radius must be greater than 0"
    
    if radius > MAX_SEARCH_RADIUS:
        return False, f"Radius must be less than {MAX_SEARCH_RADIUS} meters"
    
    return True, radius


def validate_scooter_id(scooter_id):
    """
    Validate and sanitize scooter ID
    Returns: (is_valid: bool, result: str or error_message: str)
    """
    if not scooter_id:
        return False, "Scooter ID cannot be empty"
    
    # Security: Must be a simple string, not a dict/list (injection prevention)
    if isinstance(scooter_id, (dict, list)):
        return False, "Invalid scooter ID format (possible injection attempt)"
    
    # Convert to string and strip whitespace
    scooter_id = str(scooter_id).strip()
    
    if not scooter_id:
        return False, "Scooter ID cannot be empty or whitespace only"
    
    if len(scooter_id) > MAX_SCOOTER_ID_LENGTH:
        return False, f"Scooter ID is too long (max {MAX_SCOOTER_ID_LENGTH} characters)"
    
    # Security: Check for injection patterns
    is_safe, result = sanitize_string(scooter_id, "Scooter ID")
    if not is_safe:
        return False, result
    
    # Only allow alphanumeric, dashes, and underscores
    if not re.match(r'^[a-zA-Z0-9_-]+$', scooter_id):
        return False, "Scooter ID can only contain letters, numbers, dashes, and underscores"
    
    return True, scooter_id


def validate_email(email):
    """
    Validate and sanitize email address
    Returns: (is_valid: bool, result: str or error_message: str)
    """
    if not email:
        return False, "Email cannot be empty"
    
    # Security: Must be a string
    if isinstance(email, (dict, list)):
        return False, "Invalid email format (possible injection attempt)"
    
    email = str(email).strip().lower()
    
    # Security: Check for injection patterns
    is_safe, result = sanitize_string(email, "Email")
    if not is_safe:
        return False, result
    
    # Basic format validation
    if '@' not in email or '.' not in email:
        return False, "Invalid email format"
    
    # More strict email validation
    email_pattern = re.compile(r'^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$')
    if not email_pattern.match(email):
        return False, "Invalid email format"
    
    # Check length
    if len(email) > 254:  # RFC 5321
        return False, "Email address is too long"
    
    return True, email


def validate_password(password):
    """
    Validate password strength
    Returns: (is_valid: bool, result: str or error_message: str)
    """
    if not password:
        return False, "Password cannot be empty"
    
    # Security: Must be a string
    if isinstance(password, (dict, list)):
        return False, "Invalid password format"
    
    password = str(password)
    
    if len(password) < MIN_PASSWORD_LENGTH:
        return False, f"Password must be at least {MIN_PASSWORD_LENGTH} characters"
    
    if len(password) > 128:
        return False, "Password is too long (max 128 characters)"
    
    return True, password


def validate_request_json(data, allowed_fields=None):
    """
    Validate and sanitize an entire JSON request body
    
    Args:
        data: The parsed JSON data (dict)
        allowed_fields: Optional list of allowed field names
    
    Returns: (is_valid: bool, sanitized_data: dict or error_message: str)
    """
    if data is None:
        return False, "Request body is required"
    
    if not isinstance(data, dict):
        return False, "Request body must be a JSON object"
    
    # Check for MongoDB operators in keys
    for key in data.keys():
        if isinstance(key, str) and key.startswith('$'):
            return False, f"Invalid field name: {key} ($ prefix not allowed)"
    
    # If allowed_fields specified, check for unknown fields
    if allowed_fields:
        unknown_fields = set(data.keys()) - set(allowed_fields)
        if unknown_fields:
            return False, f"Unknown fields: {', '.join(unknown_fields)}"
    
    # Recursively sanitize all values
    is_safe, result = sanitize_input(data, "request body")
    if not is_safe:
        return False, result
    
    return True, data


def validate_required_fields(data, required_fields):
    """
    Check if all required fields are present in data
    Returns: (is_valid: bool, missing_fields: list)
    """
    if not data:
        return False, required_fields
    
    missing = [f for f in required_fields if f not in data or not data[f]]
    return len(missing) == 0, missing

File: static/js/validators.js
/**
 * Frontend Validation Utilities
 * Client-side validation for coordinates and other inputs
 */

// US Geographic Bounds
const US_BOUNDS = {
    continental: {
        latMin: 24.396308,
        latMax: 49.384358,
        lngMin: -125.0,
        lngMax: -66.93457
    },
    hawaii: {
        latMin: 18.5,
        latMax: 22.5,
        lngMin: -160.5,
        lngMax: -154.5
    },
    alaska: {
        latMin: 51.0,
        latMax: 71.5,
        lngMin: -180.0,
        lngMax: -129.0
    },
    puertoRico: {
        latMin: 17.5,
        latMax: 18.6,
        lngMin: -68.0,
        lngMax: -64.5
    }
};

/**
 * Validate coordinate values
 * @param {string|number} lat - Latitude value
 * @param {string|number} lng - Longitude value
 * @param {Object} options - Validation options
 * @returns {Object} { valid: boolean, lat: number, lng: number, error: string, suggestions: string[] }
 */
function validateCoordinates(lat, lng, options = {}) {
    const { checkUSBounds = false, allowNullIsland = false } = options;
    const result = { valid: false, lat: null, lng: null, error: null, suggestions: [] };
    
    // Check for empty values
    if (lat === null || lat === undefined || lat === '') {
        result.error = 'Latitude is required';
        result.suggestions.push('Enter a latitude value (e.g., 30.2672 for Austin, TX)');
        return result;
    }
    
    if (lng === null || lng === undefined || lng === '') {
        result.error = 'Longitude is required';
        result.suggestions.push('Enter a longitude value (e.g., -97.7431 for Austin, TX)');
        return result;
    }
    
    // Convert to numbers
    const latNum = parseFloat(lat);
    const lngNum = parseFloat(lng);
    
    // Check if valid numbers
    if (isNaN(latNum)) {
        result.error = 'Latitude must be a valid number';
        result.suggestions.push('Remove any letters or special characters from latitude');
        return result;
    }
    
    if (isNaN(lngNum)) {
        result.error = 'Longitude must be a valid number';
        result.suggestions.push('Remove any letters or special characters from longitude');
        return result;
    }
    
    // Check for Infinity
    if (!isFinite(latNum) || !isFinite(lngNum)) {
        result.error = 'Coordinates must be finite numbers';
        return result;
    }
    
    // Range validation
    if (latNum < -90 || latNum > 90) {
        result.error = `Latitude must be between -90 and 90 (got ${latNum})`;
        if (lngNum >= -90 && lngNum <= 90 && (latNum >= -180 && latNum <= 180)) {
            result.suggestions.push('Latitude and longitude may be swapped. Try switching them.');
        }
        return result;
    }
    
    if (lngNum < -180 || lngNum > 180) {
        result.error = `Longitude must be between -180 and 180 (got ${lngNum})`;
        return result;
    }
    
    // Null Island check
    if (!allowNullIsland && latNum === 0 && lngNum === 0) {
        result.error = 'Coordinates (0, 0) is "Null Island" - not a valid location';
        result.suggestions.push('Enter the actual coordinates for your location');
        return result;
    }
    
    // Check for positive longitude in US (common mistake)
    if (lngNum > 0 && lngNum >= 66 && lngNum <= 125) {
        result.error = `US longitudes are negative. Did you mean -${lngNum}?`;
        result.suggestions.push(`Try using -${lngNum} for the longitude`);
        return result;
    }
    
    // US bounds check (optional)
    if (checkUSBounds) {
        const inUS = isInUSBounds(latNum, lngNum);
        if (!inUS) {
            result.error = `Location (${latNum.toFixed(4)}, ${lngNum.toFixed(4)}) appears to be outside US territory`;
            result.suggestions.push('Verify the coordinates are correct');
            result.suggestions.push('Scooters can only be placed within US locations');
            return result;
        }
    }
    
    // Valid!
    result.valid = true;
    result.lat = Math.round(latNum * 1000000) / 1000000;  // 6 decimal precision
    result.lng = Math.round(lngNum * 1000000) / 1000000;
    return result;
}

/**
 * Check if coordinates are within US territory
 */
function isInUSBounds(lat, lng) {
    // Continental US
    if (lat >= US_BOUNDS.continental.latMin && lat <= US_BOUNDS.continental.latMax &&
        lng >= US_BOUNDS.continental.lngMin && lng <= US_BOUNDS.continental.lngMax) {
        return true;
    }
    
    // Hawaii
    if (lat >= US_BOUNDS.hawaii.latMin && lat <= US_BOUNDS.hawaii.latMax &&
        lng >= US_BOUNDS.hawaii.lngMin && lng <= US_BOUNDS.hawaii.lngMax) {
        return true;
    }
    
    // Alaska (main)
    if (lat >= US_BOUNDS.alaska.latMin && lat <= US_BOUNDS.alaska.latMax &&
        lng >= US_BOUNDS.alaska.lngMin && lng <= US_BOUNDS.alaska.lngMax) {
        return true;
    }
    
    // Puerto Rico
    if (lat >= US_BOUNDS.puertoRico.latMin && lat <= US_BOUNDS.puertoRico.latMax &&
        lng >= US_BOUNDS.puertoRico.lngMin && lng <= US_BOUNDS.puertoRico.lngMax) {
        return true;
    }
    
    return false;
}

/**
 * Format coordinates for display
 */
function formatCoordinates(lat, lng, precision = 4) {
    if (lat === null || lng === null) return 'Unknown location';
    return `${parseFloat(lat).toFixed(precision)}, ${parseFloat(lng).toFixed(precision)}`;
}

/**
 * Validate a radius value
 */
function validateRadius(radius, maxRadius = 50000) {
    const result = { valid: false, radius: null, error: null };
    
    if (radius === null || radius === undefined || radius === '') {
        result.error = 'Radius is required';
        return result;
    }
    
    const radiusNum = parseFloat(radius);
    
    if (isNaN(radiusNum)) {
        result.error = 'Radius must be a valid number';
        return result;
    }
    
    if (radiusNum <= 0) {
        result.error = 'Radius must be greater than 0';
        return result;
    }
    
    if (radiusNum > maxRadius) {
        result.error = `Radius cannot exceed ${maxRadius} meters (${(maxRadius/1000).toFixed(1)} km)`;
        return result;
    }
    
    result.valid = true;
    result.radius = radiusNum;
    return result;
}

/**
 * Show validation error with visual feedback
 */
function showValidationError(inputElement, message, suggestions = []) {
    // Add error styling
    inputElement.classList.add('validation-error');
    inputElement.classList.remove('validation-success');
    
    // Find or create error message element
    let errorEl = inputElement.parentElement.querySelector('.validation-message');
    if (!errorEl) {
        errorEl = document.createElement('div');
        errorEl.className = 'validation-message';
        inputElement.parentElement.appendChild(errorEl);
    }
    
    let html = `<span class="error-text">❌ ${message}</span>`;
    if (suggestions.length > 0) {
        html += '<ul class="suggestions">';
        suggestions.forEach(s => html += `<li>💡 ${s}</li>`);
        html += '</ul>';
    }
    
    errorEl.innerHTML = html;
    errorEl.style.display = 'block';
}

/**
 * Clear validation error
 */
function clearValidationError(inputElement) {
    inputElement.classList.remove('validation-error');
    inputElement.classList.add('validation-success');
    
    const errorEl = inputElement.parentElement.querySelector('.validation-message');
    if (errorEl) {
        errorEl.style.display = 'none';
    }
}

/**
 * Clear all validation styling
 */
function clearValidation(inputElement) {
    inputElement.classList.remove('validation-error', 'validation-success');
    
    const errorEl = inputElement.parentElement.querySelector('.validation-message');
    if (errorEl) {
        errorEl.style.display = 'none';
    }
}

// Export for use in other modules
window.validateCoordinates = validateCoordinates;
window.validateRadius = validateRadius;
window.isInUSBounds = isInUSBounds;
window.formatCoordinates = formatCoordinates;
window.showValidationError = showValidationError;
window.clearValidationError = clearValidationError;
window.clearValidation = clearValidation;

